### PR TITLE
BMS 4154 BMS 4084 Import Germplasm Updates

### PR DESCRIPTION
**Pull request by abatac**

### CODE REVIEW

File: src/main/java/org/ibp/api/java/impl/middleware/common/validator/GermplasmUpdateValidator.java [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+	private org.generationcp.middleware.api.germplasm.GermplasmService germplasmMiddlewareService;
+
+	@Autowired
+	private GermplasmDataManager germplasmDataManager;
+
+	@Autowired
+	private LocationDataManager locationDataManager;
+
+	@Autowired
+	private BreedingMethodService breedingMethodService;
```
Comment (darla-leafnode) [inline src/main/java/org/ibp/api/java/impl/middleware/common/validator/GermplasmUpdateValidator.java:42]: `breedingMethodService `is unused

File: src/main/java/org/ibp/api/java/impl/middleware/common/validator/GermplasmUpdateValidator.java [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+
+	public void validateEmptyList(final BindingResult errors, final List<GermplasmUpdateDTO> germplasmUpdateDTOList) {
+		if (germplasmUpdateDTOList == null || germplasmUpdateDTOList.isEmpty()) {
+			errors.reject("germplasm.update.empty.list", "");
+		}
+	}
+
+	public void validateCodes(final BindingResult errors, final List<GermplasmUpdateDTO> germplasmUpdateDTOList) {
```
Comment (darla-leafnode) [inline src/main/java/org/ibp/api/java/impl/middleware/common/validator/GermplasmUpdateValidator.java:50]: Better to rename as `validateAttributeAndNameCodes`

File: src/main/java/org/ibp/api/java/impl/middleware/common/validator/GermplasmUpdateValidator.java [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+
+		final Set<String> attributesCode = new HashSet<>();
+		germplasmUpdateDTOList.stream().filter(germ -> germ.getAttributes() != null).collect(Collectors.toList())
+			.forEach(
+				g -> attributesCode.addAll(g.getAttributes().keySet().stream().map(n -> n.toUpperCase()).collect(Collectors.toList())));
+		final List<String> existingAttributesCode =
+			this.germplasmService.filterGermplasmAttributes(attributesCode).stream().map(AttributeDTO::getCode).collect(
+				Collectors.toList());
+
+		nameCodes.removeAll(existingNamesCode);
```
Comment (darla-leafnode) [inline src/main/java/org/ibp/api/java/impl/middleware/common/validator/GermplasmUpdateValidator.java:72]: Is it sufficient to do `nameCodes.equals(existingNamesCodes)` and `attributesCode .equals(existingAttributesCodes) `

File: src/main/java/org/ibp/api/java/impl/middleware/common/validator/GermplasmUpdateValidator.java [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+
+		final Set<String> nameCodes = new HashSet<>();
+		germplasmUpdateDTOList
+			.forEach(g -> nameCodes.addAll(g.getNames().keySet().stream().map(n -> n.toUpperCase()).collect(Collectors.toList())));
+		nameCodes.addAll(
+			germplasmUpdateDTOList.stream()
+				.map(o -> StringUtils.isNotEmpty(o.getPreferredName()) ? o.getPreferredName().toUpperCase() : null)
+				.filter(Objects::nonNull).collect(
+				Collectors.toSet()));
+		final List<String> existingNamesCode =
```
Comment (darla-leafnode) [inline src/main/java/org/ibp/api/java/impl/middleware/common/validator/GermplasmUpdateValidator.java:60]: Rename as `existingNameCodes`

File: src/main/java/org/ibp/api/java/impl/middleware/common/validator/GermplasmUpdateValidator.java [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+			this.germplasmService.filterGermplasmNameTypes(nameCodes).stream().map(GermplasmNameTypeDTO::getCode).collect(
+				Collectors.toList());
+
+		final Set<String> attributesCode = new HashSet<>();
+		germplasmUpdateDTOList.stream().filter(germ -> germ.getAttributes() != null).collect(Collectors.toList())
+			.forEach(
+				g -> attributesCode.addAll(g.getAttributes().keySet().stream().map(n -> n.toUpperCase()).collect(Collectors.toList())));
+		final List<String> existingAttributesCode =
```
Comment (darla-leafnode) [inline src/main/java/org/ibp/api/java/impl/middleware/common/validator/GermplasmUpdateValidator.java:68]: Rename as `existingAttributeCodes`

File: src/main/java/org/ibp/api/java/impl/middleware/common/validator/GermplasmUpdateValidator.java [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+		}
+		if (!attributesCode.isEmpty()) {
+			errors.reject("germplasm.update.invalid.attribute.code", new String[] {String.join(",", attributesCode)}, "");
+		}
+
+	}
+
+	public void validateGermplasmIdAndGermplasmUUID(final BindingResult errors, final List<GermplasmUpdateDTO> germplasmUpdateDTOList) {
+		// Find rows (GermplasmUpdateDTO) with blank GID and UUID.
+		final Optional<GermplasmUpdateDTO> optionalGermplasmUpdateDTO =
```
Comment (darla-leafnode) [inline src/main/java/org/ibp/api/java/impl/middleware/common/validator/GermplasmUpdateValidator.java:86]: Rename as `germplasmWithNoIdentifier`

File: src/main/java/org/ibp/api/java/impl/middleware/common/validator/GermplasmUpdateValidator.java [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+		}
+
+		final Set<Integer> gids =
+			germplasmUpdateDTOList.stream().map(dto -> dto.getGid()).filter(Objects::nonNull).collect(Collectors.toSet());
+		final Set<String> germplasmUUIDs =
+			germplasmUpdateDTOList.stream().map(dto -> StringUtils.isNotEmpty(dto.getGermplasmUUID()) ? dto.getGermplasmUUID() : null)
+				.filter(
+					Objects::nonNull).collect(Collectors.toSet());
+
+		final List<Germplasm> germplasmByGIDs = this.germplasmDataManager.getGermplasms(new ArrayList<>(gids));
```
Comment (darla-leafnode) [inline src/main/java/org/ibp/api/java/impl/middleware/common/validator/GermplasmUpdateValidator.java:99]: I suggest we implement` GermplasmService#getGermplasmByGIDs` and so as to reduce usage and dependency on `GermplasmDataManager` in BMSAPI

File: src/main/java/org/ibp/api/java/impl/middleware/common/validator/GermplasmUpdateValidator.java [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+		gids.removeAll(germplasmByGIDs.stream().map(dto -> dto.getGid()).collect(Collectors.toSet()));
+		germplasmUUIDs.removeAll(germplasmByUUIDs.stream().map(dto -> dto.getGermplasmUUID()).collect(Collectors.toSet()));
+
+		if (!gids.isEmpty()) {
+			errors.reject("germplasm.update.invalid.gid", new String[] {
+				String.join(",", gids.stream().map(o -> String.valueOf(o)).collect(
+					Collectors.toSet()))}, "");
+		}
+
+		if (!germplasmUUIDs.isEmpty()) {
```
Comment (darla-leafnode) [inline src/main/java/org/ibp/api/java/impl/middleware/common/validator/GermplasmUpdateValidator.java:111]: Can be simplified as (without removeAll):
`if (!germplasmUUIDs.equals(germplasmByGIDs.stream().map(Germplasm::getGermplasmUUID).collect(Collectors.toSet()))`

File: src/main/java/org/ibp/api/java/impl/middleware/common/validator/GermplasmUpdateValidator.java [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+
+		locationAbbrs.removeAll(abbreviations);
+
+		if (!locationAbbrs.isEmpty()) {
+			errors.reject("germplasm.update.invalid.location.abbreviation", new String[] {String.join(",", locationAbbrs)}, "");
+		}
+
+	}
+
+	public void validateBreedingMethod(final BindingResult errors, final String programUUID,
```
Comment (darla-leafnode) [inline src/main/java/org/ibp/api/java/impl/middleware/common/validator/GermplasmUpdateValidator.java:136]: `errors `and `programUUID `are unused

File: src/main/java/org/ibp/api/java/impl/middleware/germplasm/GermplasmExcelTemplateExportServiceImpl.java [hunk 10]
```diff
@@ -114,27 +129,27 @@
 	private HSSFWorkbook wb;
 
 	@Override
-	public File export(final String cropName, final String programUUID) {
+	public File export(final String cropName, final String programUUID, final boolean updateFormat) {
```
Comment (darla-leafnode) [inline src/main/java/org/ibp/api/java/impl/middleware/germplasm/GermplasmExcelTemplateExportServiceImpl.java:132]: Rename as `isGermplasmUpdateFormat`

File: src/main/java/org/ibp/api/java/impl/middleware/germplasm/GermplasmExcelTemplateExportServiceImpl.java [hunk 11]
```diff
@@ -144,93 +159,27 @@ private File generateTemplateFile(final String fileNamePath, final String cropNa
...
-		observationSheet.setColumnWidth(GermplasmExcelTemplateExportServiceImpl.OBSERVATION_SHEET_GUID_COLUMN_INDEX, 13 * 250);
+		final HSSFRow row = observationSheet.createRow(0);
+
+		final List<String> headers = new ArrayList<>(isUpdateFormat ? IMPORT_HEADERS_FOR_UPDATE : IMPORT_HEADERS);
+		final ListIterator<String> iterator = headers.listIterator();
+		while (iterator.hasNext()) {
+			final int index = iterator.nextIndex();
+			final String header = iterator.next();
+			HSSFCell cell = row.createCell(index, CellType.STRING);
+			cell.setCellStyle(this.sheetStylesMap.get(ExcelCellStyle.HEADING_STYLE_YELLOW));
```
Comment (darla-leafnode) [inline src/main/java/org/ibp/api/java/impl/middleware/germplasm/GermplasmExcelTemplateExportServiceImpl.java:179]: You are setting all cell colors as yellow. However in the import germplasm, some cells are blue, pale blue, orange. 
Is this cleared with @clarysabel ?

File: src/main/java/org/ibp/api/rest/germplasm/GermplasmResource.java [hunk 12]
```diff
@@ -161,10 +162,11 @@ public long getFilteredCount() {
 
 	@RequestMapping(value = "/crops/{cropName}/germplasm/templates/xls", method = RequestMethod.GET)
 	public ResponseEntity<FileSystemResource> getImportGermplasmExcelTemplate(@PathVariable final String cropName,
-		@RequestParam(required = false) final String programUUID) {
+		@RequestParam(required = false) final String programUUID,
+		@RequestParam(required = false) final boolean updateFormat) {
```
Comment (darla-leafnode) [inline src/main/java/org/ibp/api/rest/germplasm/GermplasmResource.java:166]: Rename as `isGermplasmUpdateFormat`

File: src/main/java/org/ibp/api/rest/germplasm/GermplasmResource.java [hunk 13]
```diff
@@ -190,6 +192,16 @@ public long getFilteredCount() {
...
 	}
 
+	@ApiOperation(value = "Import germplasm updates. Updating Breeding Method is not yet supported.")
+	@PreAuthorize("hasAnyAuthority('ADMIN', 'CROP_MANAGEMENT', 'GERMPLASM', 'MANAGE_GERMPLASM', 'IMPORT_GERMPLASM_UPDATES')")
+	@RequestMapping(value = "/crops/{cropName}/germplasm", method = RequestMethod.PATCH)
+	@ResponseBody
+	public ResponseEntity<Set<Integer>> importGermplasmUpdates(@PathVariable final String cropName,
+		@RequestParam(required = false) final String programUUID,
+		@RequestBody final List<GermplasmUpdateDTO> germplasmList) {
+		return new ResponseEntity<>(this.germplasmService.importGermplasmUpdates(programUUID, germplasmList), HttpStatus.OK);
```
Comment (darla-leafnode) [inline src/main/java/org/ibp/api/rest/germplasm/GermplasmResource.java:202]: No need to pass `programUUID`. You can use `ContextHolder.getCurrentProgram()` in `GermplasmUpdateValidator`

File: src/main/java/org/ibp/api/java/impl/middleware/germplasm/GermplasmExcelTemplateExportServiceImpl.java (hunk 11 above)
Comment (abatac) [reply src/main/java/org/ibp/api/java/impl/middleware/germplasm/GermplasmExcelTemplateExportServiceImpl.java:179]: I didn't notice that some of the headers have different colors. I'll just revert it back.

Comment (clarysabel) [reply src/main/java/org/ibp/api/java/impl/middleware/germplasm/GermplasmExcelTemplateExportServiceImpl.java:179]: You can compare with the ones in the import germplasm template. We have already reviewed the colors with Mariano using that file

File: src/main/java/org/ibp/api/rest/germplasm/GermplasmResource.java (hunk 12 above)
Comment (clarysabel) [reply src/main/java/org/ibp/api/rest/germplasm/GermplasmResource.java:166]: Are we using the same recourse for both templates? seems a bit weird to me. If that's the case, then we can add a PathVariable and witch using the path variable

File: src/main/java/org/ibp/api/java/germplasm/GermplasmService.java [hunk 14]
```diff
@@ -49,6 +50,8 @@
 
 	long countAttributesByGid(String gid, List<String> attributeDbIds);
 
+	Set<Integer> importGermplasmUpdates(String programUUID, List<GermplasmUpdateDTO> germplasmUpdateDTOList);
```
Comment (darla-leafnode) [inline src/main/java/org/ibp/api/java/germplasm/GermplasmService.java:53]: Exclude programUUID parameter from services

**Review by darla-leafnode — APPROVED** [review]

**Pull request merged**